statbot/crawler.py

```diff
@@ -11,165 +11,200 @@
+        await self.init()
 
         yield_delay = self.config['crawler']['delays']['yield']
         long_delay = self.config['crawler']['delays']['empty-source']
 
+        done = dict.fromkeys(self.progress.keys(), False)
         while True:
-            # tuple() is necessary since the underlying
-            # dict of channels may change size during
-            # an iteration
-            all_empty = True
-            for cid in tuple(self.progress.keys()):
-                # Do round-robin between all the channels
-                try:
-                    channel = self.channels[cid]
-                    mhist = self.progress[cid]
-                    all_empty &= not await self._read(channel, mhist)
-                except Exception:
-                    self.logger.error(f"Error reading (or syncing) messages from channel id {cid}", exc_info=1)
-
-            # Sleep before next cycle
-            if all_empty:
-                self.logger.info("All channels are exhausted, sleeping for a while...")
+            # Round-robin between all sources:
+            # Tuple because the underlying dictionary may change size
+            for source, last_id in tuple(self.progress.items()):
+                if done[source]:
+                    continue
+
+                events = await self.read(source, last_id)
+                if events is None:
+                    # This source is exhausted
+                    done[source] = True
+                    await self.queue.put((source, None, NOW_ID))
+                    self.progress[source] = NOW_ID
+                else:
+                    # This source still has more
+                    last_id = self.get_last_id(events)
+                    await self.queue.put((source, events, last_id))
+                    self.progress[source] = last_id
+
+            if all(done.values()):
+                self.logger.info(f"{self.name}: all sources are exhausted, sleeping for a while...")
                 delay = long_delay
             else:
                 delay = yield_delay
             await asyncio.sleep(delay)
 
-    async def _read(self, channel, mhist):
-        start_id = mhist.find_first_hole(NOW_ID)
-        if start_id is None:
-            # No more messages in this channel
-            return False
+    async def consumer(self):
+        self.logger.info(f"{self.name}: consumer coroutine started!")
 
-        start = discord.utils.snowflake_time(start_id)
+        while True:
+            source, events, last_id = await self.queue.get()
+            self.logger.info(f"{self.name}: got group of events from queue")
+
+            try:
+                with self.sql.transaction() as trans:
+                    if events is not None:
+                        await self.write(trans, events)
+                    await self.update(trans, source, last_id)
+            except SQLAlchemyError:
+                self.logger.error(f"{self.name}: error during event write", exc_info=1)
```

Comment:
I'm still not a huge fan of this since we don't have a specific idea of the conditions that caused this and whether or not it's safe to continue, nor will we likely notice because the bot will just keep running, chundering logs into the ether, but it's better than a bare except I suppose